lib/spack/spack/url_buildcache.py

```diff
@@ -0,0 +1,1129 @@
+        raise BuildcacheEntryError("v2 buildcache entries do not have a manifest file")
+
+    def remove(self):
+        raise BuildcacheEntryError("Spack cannot delete v2 buildcache entries")
+
+    def push(
+        self,
+        spec: spack.spec.Spec,
+        tarball_path: str,
+        checksum_algorithm: str,
+        tarball_checksum: str,
+        tmpdir: str,
+        signing_key: Optional[str],
+    ) -> None:
+        raise BuildcacheEntryError("Spack can no longer push v2 buildcache entries")
+
+    def destroy(self):
+        if self.archive_stage:
+            self.archive_stage.destroy()
+            self.archive_stage = None
+        if self.spec_stage:
+            self.spec_stage.destroy()
+            self.spec_stage = None
+
+
+def get_url_buildcache_class(
+    layout_version: int = CURRENT_BUILD_CACHE_LAYOUT_VERSION,
+) -> Type[URLBuildcacheEntry]:
+    if layout_version == 2:
+        return URLBuildcacheEntryV2
+    elif layout_version == 3:
+        return URLBuildcacheEntry
+    else:
+        raise UnknownBuildcacheLayoutError(
+            f"Cannot create buildcache class for unknown layout version {layout_version}"
+        )
+
+
+def check_mirrors_for_layout():
+    """Check configured mirrors, warning about any that are missing layout.json"""
+    if not config.get("config:check_mirrors_on_startup"):
+        return
+
+    cache_class = get_url_buildcache_class()
+    for mirror in MirrorCollection(binary=True).values():
+        if not cache_class.check_layout_json_exists(mirror.fetch_url):
+            msg = (
+                f"Configured mirror {mirror.name} is missing layout.json and has either \n"
+                "    never been pushed or is of an old layout version. Consider running \n"
+                "    'spack buildcache migrate' or rebuilding the specs in this mirror."
+            )
+            tty.warn(msg)
+
+
+def validate_checksum(file_path, checksum_algorithm, expected_checksum) -> None:
+    """Compute the checksum of the given file and raise if invalid"""
+    local_checksum = spack.util.crypto.checksum(
+        spack.util.crypto.hash_fun_for_algo(checksum_algorithm), file_path
+    )
+
+    if local_checksum != expected_checksum:
+        size, contents = fsys.filesummary(file_path)
+        raise spack.error.NoChecksumException(
+            file_path, size, contents, checksum_algorithm, expected_checksum, local_checksum
+        )
+
+
+def gzip_compressor(data: bytes) -> bytes:
+    buf = io.BytesIO()
+    with gzip.GzipFile(filename="", fileobj=buf, mode="wb", compresslevel=6, mtime=0) as fd:
+        fd.write(data)
+    return buf.getvalue()
+
+
+def get_compressor(compression: str = "none") -> Callable[[bytes], bytes]:
+    if compression == "none":
+        return lambda b: b
+    elif compression == "gzip":
+        return gzip_compressor
+    else:
+        raise BuildcacheEntryError(f"Unknown compression type: {compression}")
+
+
+def compress_and_write_data(
```

Comment:
In `spack.util.archive` this is a single pass, roughly `GzipFile(ChecksumWriter(open(out, "wb")))`. Could do the same here.